### DISCUSSION TITLE
[Question] prune_col in  optimizer.optimize

### DISCUSSION THREAD

**Discussion by chowc** [risinglightdb/risinglight#677] [General]
In `optimze` function: required columns are determined by `plan.out_types()`,

https://github.com/risinglightdb/risinglight/blob/c859dffc379f8ce78737d60f0e95624766cd0a2b/src/optimizer/mod.rs#L51-L52

but in `PlanNode.prune_col` function, the input types are also determined by `out_types()` of the same plan. 

**Will `required_cols` parameter always same as `input_types`?**

https://github.com/risinglightdb/risinglight/blob/c859dffc379f8ce78737d60f0e95624766cd0a2b/src/optimizer/plan_nodes/mod.rs#L155-L178

**Comment by skyzh** [comment]
@lokax might have some ideas. I'll also take a look after I've settled down recently :)
